### PR TITLE
Update the internal authentication flow of the client

### PR DESCRIPTION
**Pull request by abhipsaMisra**
We had a lot of "configuration" type classes that held client information (credentials, transport settings, options etc), and they were all inter-related in a messy way.
This PR refactors it so that we have clear separation between the concepts.

We'll pass in the pipeline context to our transport layer.
The pipeline context has the following properties:
- connection credentials (contains the required params to authenticate a client - sas tokens/ certificate information).
- product information
- model Id
- transport settings (the transport settings still holds certificate information, but I'll refactor that out in a separate PR)
- all the callbacks that a user sets on the client

This PR also refactors out the parsing logic for a device connection string and shared access signature tokens.

### CODE REVIEW

File: e2e/test/helpers/IotHubConnectionCredentialsExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Text;
+using Microsoft.Azure.Devices.Client;
+
+namespace Microsoft.Azure.Devices.E2ETests.Helpers
+{
+    internal static class IotHubConnectionCredentialsExtensions
+    {
+        public static string GetIotHubConnectionString(this IotHubConnectionCredentials iotHubConnectionCredentials)
```
Comment (abhipsaMisra) [inline e2e/test/helpers/IotHubConnectionCredentialsExtensions.cs:11]: This logic is used by a single edge related test that tests edge capabilities. Moved this logic to a test helper instead of keeping it in our sourcecode.

File: e2e/test/iothub/DeviceTokenRefreshE2ETests.cs [hunk 2]
```diff
@@ -115,14 +115,11 @@ public async Task IotHubDeviceClient_CreateFromConnectionString_TokenIsRefreshed
             using var deviceDisconnected = new SemaphoreSlim(0);
 
             TestDevice testDevice = await TestDevice.GetTestDeviceAsync(Logger, DevicePrefix).ConfigureAwait(false);
+            var auth = new DeviceAuthenticationWithConnectionString(testDevice.ConnectionString, sasTokenTimeToLive, sasTokenRenewalBuffer);
```
Comment (abhipsaMisra) [inline e2e/test/iothub/DeviceTokenRefreshE2ETests.cs:118]: We now have a specific IAuthenticationMethod that accepts connection string and the sas ttl settings. The regular CreateFromConnectionString no longer accepts configurable sas ttl.

File: iothub/device/src/AdditionalClientInformation.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    // Holder for additional client properties such as product information, PnP model Id.
+    // This class does not hold any client credential information.
+    internal class AdditionalClientInformation
```
Comment (abhipsaMisra) [inline iothub/device/src/AdditionalClientInformation.cs:12]: This is a holder for additional client information that is passed on to the transport layer. Previously we were passing in the complete client options but this now exposes only the specific required information.

File: iothub/device/src/Authentication/IConnectionCredentials.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+using System;
+using System.Security.Cryptography.X509Certificates;
+using Microsoft.Azure.Devices.Client.Transport;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    /// <summary>
+    /// Interface for client connection credentials.
+    /// This has been included for our own unit testing.
```
Comment (abhipsaMisra) [inline iothub/device/src/Authentication/IConnectionCredentials.cs:12]: Purpose of this class is primarily for unit testing

File: iothub/device/src/Authentication/IotHubConnectionCredentials.cs [hunk 5]
```diff
@@ -2,30 +2,20 @@
...
 
 namespace Microsoft.Azure.Devices.Client
 {
     /// <summary>
     /// Holder for client credentials that will be used for authenticating the client with IoT hub service.
     /// </summary>
-    public sealed class IotHubConnectionCredentials
+    public sealed class IotHubConnectionCredentials : IConnectionCredentials
     {
-        private const char ValuePairDelimiter = ';';
```
Comment (abhipsaMisra) [inline iothub/device/src/Authentication/IotHubConnectionCredentials.cs:17]: Moved the parsing logic to a dedicated parser class.

File: iothub/device/src/ClientFactory.cs [hunk 6]
```diff
@@ -105,23 +105,12 @@ internal static InternalClient Create(string hostName, IAuthenticationMethod aut
             Argument.AssertNotNull(iotHubConnectionCredentials, nameof(iotHubConnectionCredentials));
             Argument.AssertNotNull(options, nameof(options));
 
-            // Clients that derive their authentication method from AuthenticationWithTokenRefresh will need to specify
```
Comment (abhipsaMisra) [inline iothub/device/src/ClientFactory.cs:108]: This flow is no longer supported

File: iothub/device/src/Edge/EdgeModuleClientFactory.cs [hunk 7]
```diff
@@ -92,13 +92,14 @@ private async Task<IotHubModuleClient> CreateInternalClientFromEnvironmentAsync(
...
                 ISignatureProvider signatureProvider = new HttpHsmSignatureProvider(edgedUri, DefaultApiVersion);
 
-                TimeSpan sasTokenTimeToLive = _options.SasTokenTimeToLive;
-                int sasTokenRenewalBuffer = _options.SasTokenRenewalBuffer;
-
-#pragma warning disable CA2000 // Dispose objects before losing scope - IDisposable ModuleAuthenticationWithHsm is disposed when the client is disposed.
-                // Since the sdk creates the instance of disposable ModuleAuthenticationWithHsm, the sdk needs to dispose it once the client is disposed.
-                var authMethod = new ModuleAuthenticationWithHsm(signatureProvider, deviceId, moduleId, generationId, sasTokenTimeToLive, sasTokenRenewalBuffer);
-#pragma warning restore CA2000 // Dispose objects before losing scope - IDisposable ModuleAuthenticationWithHsm is disposed when the client is disposed.
+                // TODO: environment variables need to be added to accept SasTokenTimeToLive and SasTokenRenewalBuffer.
```
Comment (abhipsaMisra) [inline iothub/device/src/Edge/EdgeModuleClientFactory.cs:95]: This is an open #TODO item

File: iothub/device/src/IotHubClientOptions.cs [hunk 8]
```diff
@@ -67,30 +67,6 @@ public IotHubClientOptions(IotHubClientTransportSettings transportSettings)
...
-        /// The suggested time to live value for tokens generated for SAS authenticated clients.
-        /// The <see cref="TimeSpan"/> provided should be a positive value, signifying that it is not possible to generate tokens that have already expired.
-        /// If unset the generated SAS tokens will be valid for 1 hour.
-        /// </summary>
-        /// <remarks>
-        /// This is used only for SAS token authenticated clients through either the
-        /// <see cref="IotHubDeviceClient.CreateFromConnectionString(string, IotHubClientOptions)"/> flow, the <see cref="IotHubModuleClient.CreateFromConnectionString(string, IotHubClientOptions)"/> flow
-        /// or the <see cref="IotHubModuleClient.CreateFromEnvironmentAsync(IotHubClientOptions)"/> flow.
-        /// </remarks>
-        public TimeSpan SasTokenTimeToLive { get; set; }
```
Comment (abhipsaMisra) [inline iothub/device/src/IotHubClientOptions.cs:80]: The connection string flow has an associated IAuthenticationMethod that can be used for this feature.

File: iothub/device/src/Pipeline/TransportHandlerFactory.cs [hunk 9]
```diff
@@ -12,43 +11,31 @@ internal class TransportHandlerFactory : ITransportHandlerFactory
...
-            Func<Message, Task> onDeviceMessageReceivedCallback = context.DeviceEventCallback;
+            IotHubClientTransportSettings transportSettings = context.IotHubClientTransportSettings;
 
-            if (clientConfiguration.ClientOptions.TransportSettings is IotHubClientAmqpSettings iotHubClientAmqpSettings)
+            if (transportSettings is IotHubClientAmqpSettings iotHubClientAmqpSettings)
             {
                 return new AmqpTransportHandler(
                     context,
-                    iotHubClientAmqpSettings,
-                    onMethodCallback,
```
Comment (abhipsaMisra) [inline iothub/device/src/Pipeline/TransportHandlerFactory.cs:26]: These callbacks are a part of the pipeline context which is passed into the transport layer. It is not required for us to extract these here and pass them in again.

File: iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs [hunk 10]
```diff
@@ -17,22 +17,22 @@ internal class AmqpAuthenticationRefresher : IAmqpAuthenticationRefresher
         private static readonly string[] s_accessRightsStringArray = AccessRightsHelper.AccessRightsToStringArray(AccessRights.DeviceConnect);
         private readonly Uri _amqpEndpoint;
         private readonly AmqpIotCbsLink _amqpIotCbsLink;
-        private readonly IClientConfiguration _clientConfiguration;
```
Comment (abhipsaMisra) [inline iothub/device/src/Transport/Amqp/AmqpAuthenticationRefresher.cs:20]: The amqp changes rea only to refactor the layer to use IotHubConnectionCredentials instead of ClientConfiguration.

File: iothub/device/tests/IotHubDeviceClientTests.cs [hunk 11]
```diff
@@ -127,10 +128,10 @@ public void IotHubDeviceClient_ParamsHostNameAuthMethodTransportType_Works()
             var transportSettings = new IotHubClientAmqpSettings(IotHubClientTransportProtocol.WebSocket);
             var options = new IotHubClientOptions(transportSettings);
 
-            // TODO (abmisr): IotHubClientOptions shouldn't be required in both
```
Comment (abhipsaMisra) [inline iothub/device/tests/IotHubDeviceClientTests.cs:130]: This TODO has now been addressed

File: iothub/device/src/Authentication/AuthenticationModel.cs [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+        /// For example, myHub.azure-devices.net/devices/device1.
+        /// </summary>
+        SasIndividual,
+
+        /// <summary>
+        /// This is the authentication model where the SAS tokens generated for a client are scoped to IoT hub level; for example, myHub.azure-devices.net
+        /// This is generally not as secure as X.509 certificates or individually authenticated client authentication, as it opens up 
+        /// the IoT hub service instance to vulnerabilities in case the SAS tokens are compromised.
+        /// </summary>
+        SasGrouped,
```
Comment (timtay-microsoft) [inline iothub/device/src/Authentication/AuthenticationModel.cs:27]: Have we considered removing this support for v2?

File: iothub/device/src/Authentication/IotHubConnectionCredentials.cs [hunk 13]
```diff
@@ -112,61 +117,188 @@ public IotHubConnectionCredentials(string iotHubConnectionString)
...
-            // Host name
-            if (HostName.IsNullOrWhiteSpace())
+            // IoT Hub host name
+            if (IotHubHostName.IsNullOrWhiteSpace())
+            {
+                throw new FormatException("IoT hub service hostname to connect to must be specified.");
+            }
+
+            // Gateway host name
+            if (GatewayHostName.IsNullOrWhiteSpace())
```
Comment (timtay-microsoft) [inline iothub/device/src/Authentication/IotHubConnectionCredentials.cs:299]: Why is this now mandatory? Users don't need a gateway to connect if they want to connect to IoT hub directly

File: iothub/device/src/Authentication/AuthenticationModel.cs (hunk 12 above)
Comment (abhipsaMisra) [reply iothub/device/src/Authentication/AuthenticationModel.cs:27]: David and I spoke about this briefly this morning, and although we both agree that this is not a secure mechanism for authenticating clients, we're wary of removing the support outright since people might have taken a dependency on this feature with implementing their custom gateway solutions with multiplexing. 
That said, the plan is to dissuade people from adopting this flow, and instead showing them a flow using AAD with IoT hub service to create multiple client identities and connect as them (as a part of a level 400 sample).

File: iothub/device/src/Authentication/IotHubConnectionCredentials.cs (hunk 13 above)
Comment (abhipsaMisra) [reply iothub/device/src/Authentication/IotHubConnectionCredentials.cs:299]:
```csharp
        /// <summary>
        /// The host service that this client connects to.
        /// This can either be the IoT hub name or a gateway service name.
        /// </summary>
        string GatewayHostName { get; }
`

File: iothub/device/src/Authentication/IotHubConnectionStringParser.cs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+            {
+                throw new ArgumentException($"The connection string is missing the property: {propertyName}.");
+            }
+
+            return value;
+        }
+
+        private static string GetConnectionStringOptionalValue(IDictionary<string, string> map, string propertyName)
+        {
+            map.TryGetValue(propertyName, out string value);
```
Comment (timtay-microsoft) [inline iothub/device/src/Authentication/IotHubConnectionStringParser.cs:48]: If we are ignoring the returned boolean here, can't we just make this

```csharp
return map[propertyName];
```

File: iothub/device/src/Authentication/IotHubConnectionCredentials.cs (hunk 13 above)
Comment (abhipsaMisra) [reply iothub/device/src/Authentication/IotHubConnectionCredentials.cs:299]:
```csharp
            IsUsingGateway = !iotHubConnectionString.GatewayHostName.IsNullOrWhiteSpace();
            GatewayHostName = IsUsingGateway
                ? iotHubConnectionString.GatewayHostName
                : IotHubHostName;
```

File: iothub/device/src/Authentication/IotHubConnectionStringParser.cs (hunk 14 above)
Comment (timtay-microsoft) [reply iothub/device/src/Authentication/IotHubConnectionStringParser.cs:48]: And if we do that, we won't need a helper function to do one line of code

File: iothub/device/src/Edge/EdgeModuleClientFactory.cs (hunk 7 above)
Comment (timtay-microsoft) [reply iothub/device/src/Edge/EdgeModuleClientFactory.cs:95]: Is there any reason why you want to take them as environment variables rather than in ClientOptions? Are these env vars set up for the user in Edge scenarios?

File: iothub/device/src/Authentication/IotHubConnectionStringParser.cs (hunk 14 above)
Comment (abhipsaMisra) [reply iothub/device/src/Authentication/IotHubConnectionStringParser.cs:48]: using an indexer still leaves us open to KeyNotFoundException. This will just return "null" instead.

File: iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs [hunk 15]
```diff
@@ -38,22 +38,26 @@ static AmqpTransportHandler()
 
         internal AmqpTransportHandler(
             PipelineContext context,
-            IotHubClientAmqpSettings transportSettings,
-            Func<MethodRequestInternal, Task> onMethodCallback = null,
-            Action<TwinCollection> onDesiredStatePatchReceivedCallback = null,
-            Func<string, Message, Task> onModuleMessageReceivedCallback = null,
-            Func<Message, Task> onDeviceMessageReceivedCallback = null)
+            IotHubClientAmqpSettings transportSettings)
```
Comment (timtay-microsoft) [inline iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs:41]: I love how much simpler this parameter list is now

**Review by timtay-microsoft — APPROVED** [review]

File: iothub/device/src/Authentication/IotHubConnectionCredentials.cs (hunk 13 above)
Comment (timtay-microsoft) [reply iothub/device/src/Authentication/IotHubConnectionCredentials.cs:299]: This seems a bit confusing to me. If the user didn't pass in a gatewayhostname in their connection string, why would the gatewayhostname be anything but null in our SDK?

Comment (abhipsaMisra) [reply iothub/device/src/Authentication/IotHubConnectionCredentials.cs:299]: That's fair. I think the name "gateway" is what's misleading there. We essentially need 3 fields - IotHubHostName, GatewayHostName and a "connecting-service-endpoint" that holds the value for this expression:
```c#
connecting-service-endpoint = iotHubConnectionString.GatewayHostName.IsNullOrWhiteSpace()
                ? IotHubHostName
                : iotHubConnectionString.GatewayHostName;
```

Comment (abhipsaMisra) [reply iothub/device/src/Authentication/IotHubConnectionCredentials.cs:299]: Any suggestions for "connecting-service-endpoint"?

Comment (timtay-microsoft) [reply iothub/device/src/Authentication/IotHubConnectionCredentials.cs:299]: "connecting-service-endpoint" as in the server that this client _directly_ connects to? "hostname" seems like a good enough name to me

File: iothub/device/src/Authentication/DeviceAuthenticationWithConnectionString.cs [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+        /// The default behavior is that the token will be renewed when it has 15% or less of its lifespan left.
+        ///</param>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if <paramref name="suggestedTimeToLive"/> is a negative timespan, or if
+        /// <paramref name="timeBufferPercentage"/> is outside the range 0-100.</exception>
+        public DeviceAuthenticationWithConnectionString(
+            string connectionString,
+            TimeSpan suggestedTimeToLive = default,
+            int timeBufferPercentage = default)
+        {
+            if (suggestedTimeToLive.Ticks < 0)
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/DeviceAuthenticationWithConnectionString.cs:40]: Why not `< TimeSpan.Zero`?

File: iothub/device/src/Authentication/DeviceAuthenticationWithRegistrySymmetricKey.cs [hunk 17]
```diff
@@ -12,18 +12,56 @@ namespace Microsoft.Azure.Devices.Client
...
+        private readonly int _timeBufferPercentage;
+
         private string _deviceId;
         private byte[] _key;
 
         /// <summary>
         /// Initializes a new instance of the class.
         /// </summary>
-        /// <param name="deviceId">Device identifier.</param>
+        /// <param name="deviceId">Device Identifier.</param>
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/DeviceAuthenticationWithRegistrySymmetricKey.cs:27]: Why capitalize identifier? Is it a property noun?

File: iothub/device/src/Authentication/IotHubConnectionString.cs [hunk 18]
```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Text;
+using Microsoft.Azure.Devices.Client.Extensions;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    internal class IotHubConnectionString
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/IotHubConnectionString.cs:9]: Meaningful class comments, please.

File: iothub/device/src/Authentication/IotHubConnectionStringParser.cs [hunk 19]
```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Azure.Devices.Client.Extensions;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    internal sealed class IotHubConnectionStringParser
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/IotHubConnectionStringParser.cs:10]: Did you add unit tests?

File: iothub/device/src/Authentication/ModuleAuthenticationWithSakRefresh.cs [hunk 20]
```diff
@@ -10,33 +10,36 @@ namespace Microsoft.Azure.Devices.Client
...
+            string sharedAccessKey,
+            string sharedAccessKeyName = default,
             TimeSpan sasTokenTimeToLive = default,
             int sasTokenRenewalBuffer = default)
             : base(
-                deviceId,
-                moduleId,
-                sasTokenTimeToLive,
-                sasTokenRenewalBuffer)
+                  deviceId,
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/ModuleAuthenticationWithSakRefresh.cs:24]: This was already indented 4 spaces. Why add more?

File: iothub/device/src/Edge/EdgeModuleClientFactory.cs (hunk 7 above)
Comment (drwill-ms) [reply iothub/device/src/Edge/EdgeModuleClientFactory.cs:95]: I concur. As a rule, I don't like using environment variables for client library input, aside from the Azure Edge case.

File: iothub/device/tests/Authentication/DeviceClientConnectionStringExceptionTests.cs [hunk 21]
```diff
@@ -37,7 +37,7 @@ public void DeviceClientConnectionStringDefaultScopeDefaultCredentialTypeMissing
         }
 
         [TestMethod]
-        [ExpectedException(typeof(ArgumentException))]
+        [ExpectedException(typeof(InvalidOperationException))]
```
Comment (drwill-ms) [inline iothub/device/tests/Authentication/DeviceClientConnectionStringExceptionTests.cs:40]: Is the client documented with this new exception type?

File: iothub/device/src/Authentication/DeviceAuthenticationWithConnectionString.cs (hunk 16 above)
Comment (abhipsaMisra) [reply iothub/device/src/Authentication/DeviceAuthenticationWithConnectionString.cs:40]: This will also work. I'll change it to this to keep it clear.

File: iothub/device/src/Authentication/DeviceAuthenticationWithRegistrySymmetricKey.cs (hunk 17 above)
Comment (abhipsaMisra) [reply iothub/device/src/Authentication/DeviceAuthenticationWithRegistrySymmetricKey.cs:27]: This looks like an unintentional copy-paste issue. I'll update it.

File: iothub/device/src/Authentication/IotHubConnectionString.cs (hunk 18 above)
Comment (abhipsaMisra) [reply iothub/device/src/Authentication/IotHubConnectionString.cs:9]: sure, I'll add some comments

File: iothub/device/src/Authentication/ModuleAuthenticationWithSakRefresh.cs (hunk 20 above)
Comment (abhipsaMisra) [reply iothub/device/src/Authentication/ModuleAuthenticationWithSakRefresh.cs:24]: Probably another copy-paste side-effect. I'll take another look at the spacing here.

File: iothub/device/tests/Authentication/DeviceClientConnectionStringExceptionTests.cs (hunk 21 above)
Comment (abhipsaMisra) [reply iothub/device/tests/Authentication/DeviceClientConnectionStringExceptionTests.cs:40]: Good point, I'll check to make sure.

File: iothub/device/src/Edge/EdgeModuleClientFactory.cs (hunk 7 above)
Comment (abhipsaMisra) [reply iothub/device/src/Edge/EdgeModuleClientFactory.cs:95]: The environmental variable is intended to be for Azure Edge case only. That is the only scenario where ModuleAuthenticationWithHsm is used. This is the internal IAuthenticationMethod the SDK creates for you when the user initializes a ModuleClient through the "CreateFromEnvironmentAsync" flow. 
Since we ask the user to provide all other credential params through environment variables, i.e. iot hub name, device Id, module Id, edge URI Id, etc.), it seems reasonable to me to have them pass the sas ttl through env variables as well. That said, I do mean to confirm with the Edge team before introducing any additional env variables.

File: iothub/device/src/Authentication/DeviceAuthenticationWithConnectionString.cs (hunk 16 above)
Comment (abhipsaMisra) [reply iothub/device/src/Authentication/DeviceAuthenticationWithConnectionString.cs:40]: This, and below: addressed in https://github.com/Azure/azure-iot-sdk-csharp/pull/2630/files
